### PR TITLE
Implementation of LDAP authentication mode.

### PR DESCRIPTION
**Pull request by holtgrewe**
Tested on Active Directory server with a relatively complex directory
structure.  Should work with other AD instances but also with OpenLDAP
with hopefully only configuration and no code change.

### CODE REVIEW

**Comment by juleskers** [comment]
(I'm not an MSKCC-employee, to avoid confusion).

Dear Manuel, what a wonderful case of serendipitous timing.
For work at our institute, I've just opened a an Active-Directory pull request last week:  https://github.com/cBioPortal/cbioportal/pull/117
From your code-comments (some mention AD, others LDAP, I'm not sure what you're using at your institute, both in a mixed fashion?

We use AD (with optional LDAP gateway) at my institute, and we came out with a bare minimum of configuration, and could do without a custom UserDetailsContextMapper entirely (spring provides one that works on default AD field names, see my PullRequest for details).


I just went through your commit, and I'm happy to discover we went with mostly the same changes.
One thing I could recommend for you: the cBio login-jsp page is a lot nicer to look at for users, and can be made to work quite simply with few changes. I'd encourage you to have a look at how did that: https://github.com/juleskers/cbioportal/commit/0248e12f8afc943c0f0cebe936583805d188589e
Probably it's even as simple as taking my diff, and adding ' || authenticationMethod.equals("ldap") ' where appropriate.

kind regards,
Jules Kerssemakers,
DKFZ Heidelberg

**Comment by holtgrewe** [comment]
Dear Jules,

we are using ActiveDirectory in the typical clinical Microsoft-rich setting here ;)

I'm no expert in Spring (or Spring Security) so I implemented the first thing that I could get to work in a clean fashion. The solution should work with both Active Directory and OpenLDAP.

I agree that using the original cBioPortal login page is nicer and I would be willing to perform the necessary changes to this PR if the MSK team considers this patch for merging.

**Comment by holtgrewe** [comment]
@juleskers hm, I am trying to incorporate the login.jsp changes in this branch here. However, using this leads to an infinite redirect to `login.jsp` loop.

https://github.com/holtgrewe/cbioportal/tree/feature/ldap-authentication

**Comment by juleskers** [comment]
Dear Manuel,

I'm no expert at spring either, but.my colleagues Florian and Andreas get by :-)

I had a cursory look, asnd at first glance your form-login looks ok.
I'm at home cooking dinner right now, but will do a more in-depth analysis tomorrow morning.

Have you considered giving my branch a try?
Your ad.domain is probably something like "ad.bihealth.de", and ad.url like "ldap://ad.bihealth.de"

**Comment by juleskers** [comment]
Ha, found it!

I see that in your applicationContext-security.xml, you added another URL-intercept:
"<intercept-url pattern="/**" access="ROLE_USER" />"

/** matches every single page, including /login.jsp.
Spring stops searching after the first match, so the other url-intercepts then aren't evaluated anymore.
The effect is that to see the login-page, you must we logged in first, so spring redirects you to the login page, which requires you to sign in first, so spring redirects you to the login page, which requires you ...

You don't need this extra intercept; "block everything" is already Spring's default.
The shared-config for cBio already handles allowing only the 'safe' parts, look under the shared beans config, specifically under ""
The login page is allowed in the ${url.intercepts}, which are found here: src/main/resources/spring.security.url.intercepts

I don't blame you, those three paragraphs took me 4 weeks and being bailed out by colleagues to figure out ;-)
I can really recommend the "Spring in Action" book if you want to learn more, they sure helped me more than the (IMHO opaque) Spring online documentation.


One last thing: I noticed you use tabs to indent your code. On Github and in my local editor, those render as 8-spaces, messing up the neat layout of the files.
cBio tends to use 4-spaces as indent (as do many other shared projects, since it always renders identically everywhere).
I'd advise switching your favourite editor to 4-space indenting too (though cBio forgets about this regularly as well, see my whitespace-fixing commits)

**Comment by holtgrewe** [comment]
Hi Jules,

thank you for your suggestion, now the cBioPortal login page looks much nicer.

Cheers,
Manuel

**Comment by n1zea144** [comment]
Hi Manuel, Jules.

Sorry I'm late to the party here.  I'm just catching up.  So, does the contribution that Jules made (AD support) make this pull request moot?

**Comment by holtgrewe** [comment]
Hi, we have not managed to use Jules' approach to work with our AD domain controller :\ I would thus ask you to consider this patch for inclusion nevertheless as it also allows to use arbitrary LDAP servers.

We would then resolve the merge issues.

**Comment by juleskers** [comment]
Hi too!
tl;dr: 'generic' ldap definitely adds something over my AD-specific work; worth including!

Manuel, I'm sorry to hear you didn't get it to work. Anything specific I could help out with? I'd also like my patch to be maximally useful for others.
I cannot fathom why Spring's built-in AD-module wouldn't work with your Domain Controller..
Is it a microsoft-genuine AD controller, or a 3rd-party (/open source) reimplementation? If the latter, that may mess with the defaults, I guess.

I'm not knowledgeable enough to judge if this'll work for "arbitrary" LDAP servers, but I like what I see! Definitely more flexible than my PR :-)
From the looks of it you've made it very configurable, and I saw a mention of "openLDAP", which is definitely a big plus!

I'm betting most the login.jsp will "cancel out" with my earlier PR.
Hopefully all that remains is replacing all my "authenticationMethod.equals("ad")" to "authenticationMethod.equals("ad") || authenticationMethod.equals("ldap") "

**Comment by holtgrewe** [comment]
Hi,

our hospital is running a "genuine Microsoft" installation but I don't know all the tweaks that central IT uses and boxes they checked (more than ten thousand users, spread over multiple locations all over the city).

Possibly someone with better knowledge about Spring Authentication might be able to get it to run but it did not run out of the box and after some hours I stopped trying to make Jules' patch work.

I also hope that in the end, my changes will be only additions and conflicts besides the `authenticationMethod.equals("ad")` will be few, if any.

Cheers

**Comment by morungos** [comment]
I don't know if this is of interest to you guys, but we've recently started deploying with an LDAP-backed OpenID Connect provider. This matches our long-term institutional plans, but for now, we've integrated into a single web container, without fairly minimal changes to the core cBioPortal. This was critical for us, as we have multiple LDAPs, needed authorization control yesterday, and I didn't want to write a whole login system.

We're also a Microsofty place, with some of the LDAPs being Active Directory, which did change the recommended authentication system. It's a bit improvised right now, and we've not made a PR from the config changes for OIDC, but I can write more if it'll help.

**Comment by holtgrewe** [comment]
Superseded by #1072.
